compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDelegatedPropertyChecker.kt

```diff
@@ -0,0 +1,95 @@
+                reporter.reportOn(delegateSource, FirErrors.RECURSION_IN_IMPLICIT_TYPES, context)
+                return
+            }
+        }
+
+        class Visitor(private val isGet: Boolean) : FirVisitorVoid() {
+            override fun visitElement(element: FirElement) = element.acceptChildren(this)
+
+            override fun visitFunctionCall(functionCall: FirFunctionCall) {
+                val errorNamedReference = functionCall.calleeReference as? FirErrorNamedReference ?: return
+                if (errorNamedReference.source?.kind != FirFakeSourceElementKind.DelegatedPropertyAccessor) return
+                val calleeArgTypes = "(${functionCall.arguments.joinToString(", ") { it.typeRef.coneType.render() }})"
+                when (val diagnostic = errorNamedReference.diagnostic) {
+                    is ConeUnresolvedNameError -> {
+                        reporter.reportOn(
+                            errorNamedReference.source,
+                            FirErrors.DELEGATE_SPECIAL_FUNCTION_MISSING,
+                            "${diagnostic.name}$calleeArgTypes",
+                            delegateType,
+                            if (isGet) "delegate" else "delegate for var (read-write property)",
+                            context
+                        )
+                    }
+                    is ConeAmbiguityError -> {
+                        if (diagnostic.applicability.isSuccess) {
+                            reporter.reportOn(
+                                errorNamedReference.source,
+                                FirErrors.DELEGATE_SPECIAL_FUNCTION_AMBIGUITY,
+                                "${diagnostic.name}$calleeArgTypes",
+                                diagnostic.candidates,
+                                context
+                            )
+                        } else {
+                            reporter.reportOn(
+                                errorNamedReference.source,
+                                FirErrors.DELEGATE_SPECIAL_FUNCTION_NONE_APPLICABLE,
+                                "${diagnostic.name}$calleeArgTypes",
+                                diagnostic.candidates,
+                                context
+                            )
+                        }
+                    }
+                    is ConeInapplicableCandidateError -> {
+                        reporter.reportOn(
+                            errorNamedReference.source,
+                            FirErrors.DELEGATE_SPECIAL_FUNCTION_NONE_APPLICABLE,
+                            "${errorNamedReference.name}$calleeArgTypes",
+                            listOf(diagnostic.candidate.symbol),
+                            context
+                        )
+                    }
+                }
+            }
+        }
+
+        declaration.getter?.acceptChildren(Visitor(true))
+        declaration.setter?.acceptChildren(Visitor(false))
```

Comment:
Hm, I'm not sure about this structure. Can we rather make `FirDelegatedPropertyChecker` a subtype of an expression checker and check a function call with `FirFakeSourceElementKind.DelegatedPropertyAccessor`? From the context, we can see whether it's containing declaration is a delegated property. Then, maybe we don't need `FirErrors.RECURSION_IN_IMPLICIT_TYPES` handling above.

Reply:
Could you clarify why we won't need `FirErrors.RECURSION_IN_IMPLICIT_TYPES` in that case? This error as for now is not reported anywhere. Also, if we make this checker an expression checker, then it would be harder to skip reporting if the delegate creation contains errors, as described in https://github.com/JetBrains/kotlin/pull/4287#discussion_r609081875

Reply:
Oh, I naively thought that `FirErrors.RECURSION_IN_IMPLICIT_TYPES` is an additional bail-out to avoid redundant errors on errored delegate. 🤦  So, that particular one indeed needs a property checker.

Okay, maybe, the current structure (as a property checker) is a better way to handle, e.g., suppression on delegated property, too.